### PR TITLE
Bulk Integration in the JS client

### PR DESCRIPTION
**Pull request by yleprincecraft**

### CODE REVIEW

**Comment by yleprincecraft** [comment]
**Todo:**
* Replace `timestamp` hardcoded values with `CONFIGURATION` timestamps

**Comment by yleprincecraft** [comment]
:wave: Hey Perrine !

Alt Text

File: src/client.js [hunk 1]
```diff
@@ -30,6 +30,47 @@ function isUrl(url) {
...
+
+// function testId(id) {
+//   if (isUnvalidId(id)) {
+//     throw new CraftAiBadRequestError(
+//       `Bad Request, unable to handle agent ${id} because it is an invalid agent id. It must only contain characters in 'a-zA-Z0-9_-' and must be a string between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
+//     );
+//   }
+// }
+
+function testOperations(operations, id = undefined) {
```
Comment (Giska) [inline src/client.js]: Since it is used only one time, set it into the function that call it directly

File: src/client.js [hunk 2]
```diff
@@ -30,6 +30,47 @@ function isUrl(url) {
...
+  if (areUnvalidOperations(operations)) {
+    throw new CraftAiBadRequestError(
+      `Bad Request, unable to handle operations for agent ${id}. Operations should be provided within an array.`
+    );
+  }
+}
+
+function testInputBulk(bulkArray) {
```
Comment (Giska) [inline src/client.js]: Rename it checkBulkParameters

File: src/client.js [hunk 3]
```diff
@@ -180,36 +286,118 @@ export default function createClient(tokenOrCfg) {
...
+          } operation(s) to the agent ${cfg.owner}/${
+            cfg.project
+          }/${agentId} context.`;
           debug(message);
           return { message };
         });
     },
-    getAgentContextOperations: function(agentId, start = undefined, end = undefined) {
+    addAgentContextOperationsBulk: function(agentsOperationsList) {
+      testInputBulk(agentsOperationsList);
```
Comment (Giska) [inline src/client.js]: rename it accordingly to the new name

File: src/client.js [hunk 4]
```diff
@@ -101,9 +164,23 @@ export default function createClient(tokenOrCfg) {
           return body;
         });
     },
+    createAgentBulk: function(agentsList) {
+      testInputBulk(agentsList);
```
Comment (Giska) [inline src/client.js]: ranme it accordingly to the new name

File: src/client.js [hunk 5]
```diff
@@ -292,7 +508,7 @@ export default function createClient(tokenOrCfg) {
         .then(({ body }) => {
           if (_.isUndefined(t)) {
             return body.shortUrl;
-          }
+          }
```
Comment (Giska) [inline src/client.js]: trailing whitespace

File: src/client.js [hunk 6]
```diff
@@ -305,47 +521,68 @@ export default function createClient(tokenOrCfg) {
...
+            'x-craft-ai-tree-version': version
+          }
+        })
+          .then(({ body }) => body);
 
       if (!cfg.decisionTreeRetrievalTimeout) {
         // Don't retry
         return agentDecisionTreeRequest();
-      }
+      }
```
Comment (Giska) [inline src/client.js]: trailing whitespace

File: src/client.js [hunk 7]
```diff
@@ -357,16 +594,38 @@ export default function createClient(tokenOrCfg) {
         ]);
       }
     },
+    getAgentDecisionTreeBulk: function(agentsList) {
+      testInputBulk(agentsList);
```
Comment (Giska) [inline src/client.js]: Rename it accordingly to the new name

File: src/client.js [hunk 8]
```diff
@@ -30,6 +30,47 @@ function isUrl(url) {
...
+const areUnvalidOperations = (operations) =>
+  _.isUndefined(operations) || !_.isArray(operations);
+
+// function testId(id) {
+//   if (isUnvalidId(id)) {
+//     throw new CraftAiBadRequestError(
+//       `Bad Request, unable to handle agent ${id} because it is an invalid agent id. It must only contain characters in 'a-zA-Z0-9_-' and must be a string between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
+//     );
+//   }
+// }
```
[the remark this answers is not in the document]
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js [hunk 9]
```diff
@@ -48,16 +91,20 @@ export default function createClient(tokenOrCfg) {
     cfg.owner = cfg.owner || owner;
     cfg.project = cfg.project || project;
     cfg.url = cfg.url || platform;
-  }
+  }
```
[the remark this answers is not in the document]
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js [hunk 10]
```diff
@@ -133,17 +214,36 @@ export default function createClient(tokenOrCfg) {
           return body;
         });
     },
+    deleteAgentBulk: function(agentsList) {
+      testInputBulk(agentsList);
+      // agentsList.map(({ id }) => testId(id));
```
[the remark this answers is not in the document]
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js [hunk 11]
```diff
@@ -180,36 +286,118 @@ export default function createClient(tokenOrCfg) {
...
+          }/${agentId} context.`;
           debug(message);
           return { message };
         });
     },
-    getAgentContextOperations: function(agentId, start = undefined, end = undefined) {
+    addAgentContextOperationsBulk: function(agentsOperationsList) {
+      testInputBulk(agentsOperationsList);
+      agentsOperationsList.map(({ id, operations }) => {
+        // testId(id);
```
[the remark this answers is not in the document]
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js [hunk 12]
```diff
@@ -180,36 +286,118 @@ export default function createClient(tokenOrCfg) {
...
+            currentChunkSize = 0;
+          } 
+          else {
+            currentChunkSize += agent.operations.length;
+            currentChunk.push(agent);
+          }
+        }
+      }
+
+      if (currentChunk.length) chunkedData.push(currentChunk);
```
[the remark this answers is not in the document]
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js (hunk 5 above)
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js (hunk 6 above)
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js (hunk 2 above)
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js (hunk 4 above)
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js (hunk 7 above)
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js (hunk 3 above)
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

File: src/client.js (hunk 1 above)
Comment (yleprincecraft) [reply src/client.js]: :heavy_check_mark:

**Pull request merged**